### PR TITLE
Feature/ APP-630 create plugin factory

### PR DESCRIPTION
**Pull request by Rekard0**
## Description

Please include a summary of the change and be sure you follow the contributions rules we do provide [here](./CONTRIBUTIONS.md)

Task: [ID]()

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] I have selected the correct base branch.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran all tests with success and extended them if necessary.
- [ ] I have updated the ``CHANGELOG.md`` file in the root folder.
- [ ] I have tested my code on the test network.

### CODE REVIEW

File: packages/contracts/contracts/votings/whitelist/WhitelistVotingFactory.sol [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import "../../plugin/IPluginFactory.sol";
+
+contract WhitelistVotingFactory is IPluginFactory {
+    using Address for address;
+    using Clones for address;
+
+    // TODO: Move it to MajorityVoting.
+    struct VoteConfig {
+        uint64 participationRequiredPct;
+        uint64 supportRequiredPct;
```
Comment (brickpop) [inline packages/contracts/contracts/votings/whitelist/WhitelistVotingFactory.sol:21]: Couldn't this be simpler in terms of a multisig? A single variable like `minSupportPct` seems to be able to do the job, unless we want to recycle the same struct within MajorityVoting

File: packages/contracts/deploy/05_ens.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import {ENS_ADDRESSES, getContractAddress, setupENS} from './helpers';
+
+const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
+  const {deployments, getNamedAccounts, ethers, getChainId, network} = hre;
+  const {deploy} = deployments;
+
+  const {deployer} = await getNamedAccounts();
+
+  const managingDAOAddress = await getContractAddress('DAO', hre);
+  const node = ethers.utils.namehash('dao.eth');
```
Comment (brickpop) [inline packages/contracts/deploy/05_ens.ts:12]: I think we can reuse this from an `util` source file that should be merged or about to be

File: packages/contracts/deploy/60_dao-factory.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import {DeployFunction} from 'hardhat-deploy/types';
+import {getContractAddress} from './helpers';
+
+const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
+  const {deployments, getNamedAccounts, ethers} = hre;
+  const {deploy} = deployments;
+
+  const {deployer} = await getNamedAccounts();
+
+  const managingDAOAddress = await getContractAddress('DAO', hre);
```
Comment (brickpop) [inline packages/contracts/deploy/60_dao-factory.ts:11]: How do we explicitly know that we will be getting the address of the managing DAO? Shouldn't this be a well-known contract deployed once?

File: packages/contracts/deploy/90_erc20-voting-factory.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  const tokenFactoryAddress = await getContractAddress('TokenFactory', hre);
+
+  const ret = await deploy('ERC20VotingFactory', {
+    from: deployer,
+    args: [tokenFactoryAddress],
+    log: true,
+  });
+
+  // Create Repo and register on Aragon Plugin Registry
+  const factoryAddress: string = ret.receipt?.contractAddress || '';
```
Comment (brickpop) [inline packages/contracts/deploy/90_erc20-voting-factory.ts:20]: How about this?

```suggestion
  const erc20VotingFactoryAddress: string = ret.receipt?.contractAddress || '';
```

File: packages/contracts/deploy/90_erc20-voting-factory.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+  if (factoryAddress !== '') {
```
Comment (brickpop) [inline packages/contracts/deploy/90_erc20-voting-factory.ts:22]: What's the use case here? How would be not want to deploy something? Is redeployment parameterizable? How would versioning work here?

File: packages/contracts/deploy/91_whitelist-voting-factory.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+  const {deployer} = await getNamedAccounts();
+
+  const ret = await deploy('WhitelistVotingFactory', {
+    from: deployer,
+    log: true,
+  });
+
+  // Create Repo and register on APM
+  const factoryAddress: string = ret.receipt?.contractAddress || '';
```
Comment (brickpop) [inline packages/contracts/deploy/91_whitelist-voting-factory.ts:17]: Same as above

File: packages/contracts/deploy/helpers.ts [hunk 7]
```diff
@@ -46,5 +53,80 @@ export async function updateActiveContractsJSON(payload: {
...
+    ensDomainHash('', ethers),
+    ensLabelHash('eth', ethers),
+    deployer
+  );
+
+  await ensRegistryContract.setSubnodeOwner(
+    ensDomainHash('eth', ethers),
+    ensLabelHash('dao', ethers),
+    deployer
+  );
```
Comment (brickpop) [inline packages/contracts/deploy/helpers.ts:116]: This only works for `dao.eth` (mainnet). Is it possible/reasonable to make it parameterizable to account for other chains/environments?

File: packages/contracts/test/factory/dao-factory.ts [hunk 8]
```diff
@@ -21,19 +22,19 @@ const ACLAnyAddress = '0xFFfFfFffFFfffFFfFFfFFFFFffFFFffffFfFFFfF';
...
-  const {name, dao, token, creator} = events.find(
+  const {name, dao, creator} = events.find(
     ({event}: {event: any}) => event === EVENTS.NewDAORegistered
   ).args;
 
+  const token = tokenVoting
+    ? events.find(({event}: {event: any}) => event === EVENTS.TokenCreated).args
+        .token
+    : zeroAddress;
+
```
Comment (brickpop) [inline packages/contracts/test/factory/dao-factory.ts:37]: Same as already commented, we will be supporting more than one token standard

**Review by brickpop — COMMENTED** [review]
Review part 1

File: packages/contracts/contracts/utils/UUPSProxy.sol [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+pragma solidity 0.8.10;
+
+import "@openzeppelin/contracts/proxy/ERC1967/ERC1967Proxy.sol";
+
+/// @title Proxy contract for UUPSUpgradeable
+/// @author Aragon Association - 2022
+/// @notice This contract is only meant to be used for deploy purposes
+/// TODO: Need to be fully tested, and decide to use the second address param as admin address.
+contract UUPSProxy is ERC1967Proxy {
```
[the remark this answers is not in the document]
Comment (Michael-A-Heuer) [reply packages/contracts/contracts/utils/UUPSProxy.sol:11]: The acronym UUPS stands for Universal Upgradeable Proxy Standard.

**Pull request closed without merging**